### ISSUE TITLE
Unhandled promise error on failed batch request

### ISSUE THREAD

**Issue by martinlaukkanen** [XRM-OSS/Xrm-WebApi-Client#52]
Hi,

I've found this problem which appears to be similar to: https://github.com/XRM-OSS/Xrm-WebApi-Client/issues/24, when a batch request fails with 400 or similar, this line of code in the FormatError function will have an unhandled exception:

https://github.com/XRM-OSS/Xrm-WebApi-Client/blob/4736dd01df7ded3f6d9c17d40628b604ce61645b/src/js/WebApiClient.Core.js#L244

Basically the xhr.response is actually just text, e.g.:
"--batchresponse_44ccd300-2937-445c-95b4-b4b14f23dcc2\r\nContent-Type: multipart ....."

So parsing to JSON fails but is not caught resulting in the issue.

**Comment by DigitalFlow** [comment]
Hi @martinlaukkanen,

thanks for reporting this issue.

Would it be possible for you to provide me a reproduction of this issue on a default entity with default fields?
That would help me a lot.

Kind regards,
Florian

**Comment by martinlaukkanen** [comment]
Hi,

Sure, the below generates a 400 response on the Contact entity by sending >50 chars for the firstname attribute:

```
const contact = WebApiClient.Create({
entityName : 'contact',
entity: {
	firstname: 'Lorem ipsum dolor sit amet, consectetur adipiscing nec'
},
asBatch: true,
});
const changes = [contact];

const batch = new WebApiClient.Batch({
changeSets: [new WebApiClient.ChangeSet({ requests: changes })],
});

const result = await WebApiClient.SendBatch(batch);
```

The xhr.response looks like:

```
--batchresponse_d22a95a3-78c5-4173-958f-5f74b786ac1f
Content-Type: multipart/mixed; boundary=changesetresponse_a622bf32-0e43-4575-9da5-fbb6a4c17f75

--changesetresponse_a622bf32-0e43-4575-9da5-fbb6a4c17f75
Content-Type: application/http
Content-Transfer-Encoding: binary
Content-ID: 1

HTTP/1.1 400 Bad Request
REQ_ID: 14c1965f-1aa0-48ac-bea7-55b47a85d1ed
Content-Type: application/json; odata.metadata=minimal
OData-Version: 4.0

{"error":{"code":"0x80044331","message":"A validation error occurred.  The length of the 'firstname' attribute of the 'contact' entity exceeded the maximum allowed length of '50'."}}
--changesetresponse_a622bf32-0e43-4575-9da5-fbb6a4c17f75--
--batchresponse_d22a95a3-78c5-4173-958f-5f74b786ac1f--
```

And the exception is uncaught. BTW when run synchronously the error is a JSON parse error which is what that line referenced above shows and not the batch response error.

**Comment by martinlaukkanen** [comment]
Hi.
I needed a quick fix for this one for my project, so I made the following change to the FormatError function:

`    function FormatError (xhr) {
        if (xhr && xhr.response) {
			var response = ParseResponse(xhr);

			if (response instanceof WebApiClient.BatchResponse) {
				var errors = "";
				if (response.errors.length > 0) {
					errors = response.errors.map(function(e) {
						return e.code + ": " + e.message;
					}).join("\n\r");
				}

				return xhr.status + " - " + errors;
			}

            var json = JSON.parse(xhr.response);

...
`

**Comment by DigitalFlow** [comment]
Hi @martinlaukkanen,

thanks again for your assistance, your PR was released as v4.0.7.

Sorry for taking that long for the release, the technology stack was hopelessly outdated and I did not feel comfortable releasing it with 71 vulnerabilities in dependencies, of which 38 were high and 9 critical...

The new release is on a completely new stack with webpack and jest as infrastructure, we now currently have exactly `zero` vulnerabilities again.

Again, thanks for your help :)

Kind regards,
Florian
